File: ios/ZapLivre/ZapLivre/Components/TypingIndicator.swift
//
//  TypingIndicator.swift
//  ZapLivre
//
//  Created by ZapLivre Team
//  Copyright © 2026 ZapLivre. All rights reserved.
//

import SwiftUI

/// TypingIndicator - Animated dots showing someone is typing
struct TypingIndicator: View {
    let peerName: String

    var body: some View {
        HStack(spacing: 8) {
            // Animated dots
            HStack(spacing: 4) {
                ForEach(0..<3) { index in
                    AnimatedDot(delay: Double(index) * 0.15)
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(Color(UIColor.systemGray5))
            .cornerRadius(16)

            // "está digitando..." text
            Text("\(peerName) está digitando...")
                .font(.caption)
                .foregroundColor(.secondary)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }
}

/// AnimatedDot - Single animated dot
struct AnimatedDot: View {
    let delay: Double

    @State private var isAnimating = false

    var body: some View {
        Circle()
            .fill(Color.gray)
            .frame(width: 8, height: 8)
            .opacity(isAnimating ? 1.0 : 0.3)
            .animation(
                Animation
                    .easeInOut(duration: 0.6)
                    .repeatForever(autoreverses: true)
                    .delay(delay),
                value: isAnimating
            )
            .onAppear {
                isAnimating = true
            }
    }
}

/// Compact typing indicator (just dots, no text)
struct TypingIndicatorCompact: View {
    var body: some View {
        HStack(spacing: 4) {
            ForEach(0..<3) { index in
                AnimatedDot(delay: Double(index) * 0.15)
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(Color(UIColor.systemGray5))
        .cornerRadius(16)
    }
}

#Preview {
    VStack(spacing: 20) {
        TypingIndicator(peerName: "João")

        HStack {
            TypingIndicatorCompact()
            Spacer()
        }
        .padding(.horizontal)
    }
}

File: android/app/src/main/kotlin/com/zaplivre/ui/components/TypingIndicator.kt
package com.zaplivre.ui.components

import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

/**
 * TypingIndicator - Animated dots showing someone is typing
 */
@Composable
fun TypingIndicator(
    peerName: String = "Contato",
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Animated dots
        Row(
            horizontalArrangement = Arrangement.spacedBy(4.dp),
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .background(
                    color = MaterialTheme.colorScheme.surfaceVariant,
                    shape = MaterialTheme.shapes.medium
                )
                .padding(horizontal = 12.dp, vertical = 8.dp)
        ) {
            // Three dots with staggered animation
            repeat(3) { index ->
                AnimatedDot(
                    delay = index * 150
                )
            }
        }

        Spacer(modifier = Modifier.width(8.dp))

        // "está digitando..." text
        Text(
            text = "$peerName está digitando...",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f)
        )
    }
}

/**
 * AnimatedDot - Single animated dot
 */
@Composable
private fun AnimatedDot(
    delay: Int = 0,
    color: Color = Color(0xFF90A4AE)
) {
    val infiniteTransition = rememberInfiniteTransition(label = "dot_animation")

    val alpha by infiniteTransition.animateFloat(
        initialValue = 0.3f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(
                durationMillis = 600,
                delayMillis = delay,
                easing = LinearEasing
            ),
            repeatMode = RepeatMode.Reverse
        ),
        label = "dot_alpha"
    )

    Box(
        modifier = Modifier
            .size(8.dp)
            .alpha(alpha)
            .background(color, CircleShape)
    )
}

/**
 * Compact typing indicator (just dots, no text)
 */
@Composable
fun TypingIndicatorCompact(
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .background(
                color = MaterialTheme.colorScheme.surfaceVariant,
                shape = MaterialTheme.shapes.medium
            )
            .padding(horizontal = 12.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        repeat(3) { index ->
            AnimatedDot(delay = index * 150)
        }
    }
}
